### PR TITLE
crimson/osd/pg: do_osd_ops_execute fixes

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Option 3 in https://tracker.ceph.com/issues/63844#note-8

Fixes: https://tracker.ceph.com/issues/63844
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by Matan-B — CHANGES_REQUESTED** [review]
I think  that we should increase `num_write` even when the operation couldn't actually complete as this counter is there to count the ops themselves.
For "actual" (successful) num_writes we use `delta_stats.num_wr++;`.

Perhaps we are not using the two correctly in another place?

**Comment by cyx1231st** [comment]
> https://tracker.ceph.com/issues/63844
> 1. reads/stats is concurrent with writes of the same object in SeaStore

The test looks only fails with SeaStore. I'm wondering how does the AlienStore behave differently for concurrent reads&stats&writes, so that it won't fail?

**Comment by xxhdx1985126** [comment]
> > https://tracker.ceph.com/issues/63844
> > 
> > 1. reads/stats is concurrent with writes of the same object in SeaStore
> 
> The test looks only fails with SeaStore. I'm wondering how does the AlienStore behave differently for concurrent reads&stats&writes, so that it won't fail?

I think BlueStore doesn't have this issue because when BlueStore::queue_transaction() returns the data modification is on disk and the metadata has already been recorded in its onode cache.

I met with another issue which I think was due to a similar issue in SeaStore: https://tracker.ceph.com/issues/67916.

I think maybe we can make `ReplicatedBackend` recognize the primary osd's transaction committed future as the "submitted" future, which should fix both issues. What do you think? @cyx1231st @athanatos @Matan-B

**Comment by cyx1231st** [comment]
https://tracker.ceph.com/issues/63844
> 1. ClientRequest_1(create obj1, write to off1 of obj1) is submitted to SeaStore;
> 2. ClientRequest_2(create obj1, write to off1 of obj1) arrives and get the obc
> 3. ClientRequest_3(create obj1, write to off1 of obj1) arrives and get the obc
> 4. Before ClientRequest_1 is committed, ClientRequest_2(create obj1) arrives and finds that obj1 already existed;
> 5. ClientRequest_2 is rollbacked, which means PG tries to load the obc from disk, but it'll get an empty obc since ClientRequest_1 is not committed, so the in-memory obc of obj1 is set to empty;
> 6. ClientRequest_3 gets executed and find that the obc is empty, so it tries to create obj1 again(by OP_TOUCH) and write new data to off1 of obj1

Regardless of how the ObjectStore is implemented, once submitted, looks if ClientRequest_1 makes the obc present in the registry at the same time, ClientRequest_2 and 3 will be correct in this case, right?

**Comment by xxhdx1985126** [comment]
> https://tracker.ceph.com/issues/63844
> 
> > 1. ClientRequest_1(create obj1, write to off1 of obj1) is submitted to SeaStore;
> > 2. ClientRequest_2(create obj1, write to off1 of obj1) arrives and get the obc
> > 3. ClientRequest_3(create obj1, write to off1 of obj1) arrives and get the obc
> > 4. Before ClientRequest_1 is committed, ClientRequest_2(create obj1) arrives and finds that obj1 already existed;
> > 5. ClientRequest_2 is rollbacked, which means PG tries to load the obc from disk, but it'll get an empty obc since ClientRequest_1 is not committed, so the in-memory obc of obj1 is set to empty;
> > 6. ClientRequest_3 gets executed and find that the obc is empty, so it tries to create obj1 again(by OP_TOUCH) and write new data to off1 of obj1
> 
> Regardless of how the ObjectStore is implemented, once submitted, looks if ClientRequest_1 makes the obc present in the registry at the same time, ClientRequest_2 and 3 will be correct in this case, right?

If `submitted` means only putting the IO request in SeaStore, the above step 4~6 may happen and ClientRequest_3 won't be correct; on the other hand, if `submitted` means the SeaStore journal committed, I don't think there'll be any problem.

> 
> Once submitted, ObjectStore will finally commit the change if the OSD don't break/restart. And if there is, anything in the memory will be lost as well.

**Comment by athanatos** [comment]
The problem with making submitted wait on seastore journal commit is that it limits concurrency -- the next write in the pipeline couldn't enter process until the commit happens.  If we expose the obc in the registry at submission time and remove read-obc-on-error behavior by copying obs for each write, would that avoid the problem?

**Comment by xxhdx1985126** [comment]
> The problem with making submitted wait on seastore journal commit is that it limits concurrency -- the next write in the pipeline couldn't enter process until the commit happens. If we expose the obc in the registry at submission time and remove read-obc-on-error behavior by copying obs for each write, would that avoid the problem?

Yeah, just changed the code to adopt this method.

File: src/crimson/osd/object_context.h [hunk 1]
```diff
@@ -75,6 +75,16 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
...
+  ObjectContext(const ObjectContext &other)
+    : obs(other.obs),
+      ssc(new SnapSetContext(*other.ssc))
+  {}
+
+  ObjectContext& operator=(const ObjectContext &other) {
+    obs = other.obs;
+    ssc = new SnapSetContext(*other.ssc);
+    return *this;
+  }
```
Comment (Matan-B) [inline src/crimson/osd/object_context.h:87]: Since rollbacker is the only user, should we introduce a dedicated `get_obc_rollbacker` instead of overriding operator= and a copy ctor? Allowing this may confuse future users.
```
// We may need to rollback the ObjectContext on failed op execution.
// Copy the current obc before mutating it in order to recover on failures.
ObjectContext get_obc_rollbacker (const ObjectContext &other) {
```

**Review by Matan-B — COMMENTED** [review]
lgtm.
Should we close #59615? If so let's add a "fixes" for https://tracker.ceph.com/issues/67916 as well.

File: src/crimson/osd/pg.cc [hunk 2]
```diff
@@ -929,9 +929,9 @@ PG::do_osd_ops_execute(
   FailureFunc&& failure_func)
 {
   assert(ox);
-  auto rollbacker = ox->create_rollbacker([this] (auto& obc) {
-    return obc_loader.reload_obc(obc).handle_error_interruptible(
-      load_obc_ertr::assert_all{"can't live with object state messed up"});
+  auto rollbacker = ox->create_rollbacker([object_context=*obc] (auto& obc) {
+    obc = object_context;
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc:933]: The new (copied) obc won't be in `obc_registry`, right? This can be a bit tricky for future cache usages. Perhaps we should handle this case as well? Or can we assume the the copied obc should not be cached and may be only used on the failure path.

**Review by Matan-B — COMMENTED** [review]
Can we change `rollback_obc_if_modified` to be `void` now that we don't actually reload the obc from disk. This can help with future simplification of the failure path.

**Comment by xxhdx1985126** [comment]
> ~lgtm.~ few suggestions in the next comment. Should we close #59615? If so let's add a "fixes" for https://tracker.ceph.com/issues/67916 as well.

Sorry, I think I misled the discussion. These two issues are not that close, it was just that the fixes I proposed overlaps with each other. I'll remove the "related issue" in both tickets.

File: src/crimson/osd/pg.cc (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:933]: > Or can we assume the the copied obc should not be cached and may be only used on the failure path.

Yeah, I think we can have that assumption.

**Comment by xxhdx1985126** [comment]
@Matan-B All comments addressed, please take a look again, thanks:-)

File: src/crimson/osd/ops_executer.h [hunk 3]
```diff
@@ -616,7 +616,10 @@ OpsExecuter::RollbackHelper<Func>::rollback_obc_if_modified(
     ox->obc->get_oid(),
     e,
     need_rollback);
-  return need_rollback ? func(*ox->obc) : interruptor::now();
+  if (need_rollback) {
+    func(ox->obc);
+  }
+  return interruptor::now();
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.h:622]: Can `rollback_obc_if_modified` also be voided?
We could simply call it instead of chaining when needed, right?

**Review by Matan-B — COMMENTED** [review]
rollback_obc_if_modified comment of:
```
  // Let's just reload from the store. Evicting from the shared
  // LRU would be tricky as next MOSDOp (the one at `get_obc`
  // phase) could actually already finished the lookup. Fortunately,
  // this is supposed to live on cold  paths, so performance is not
  // a concern -- simplicity wins.
```
Is not relevant any more as we aren't reloading anymore.

File: src/crimson/osd/ops_executer.h (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.h:622]: Ok, will change it, too:-)

Comment (Matan-B) [reply src/crimson/osd/ops_executer.h:622]: Thanks, I'll await with the next batch to include this one too.

**Comment by xxhdx1985126** [comment]
@Matan-B Done:-)

**Review by Matan-B — APPROVED** [review]
lgtm, we should refactor `do_osd_ops_execute` after merging the relevant changes there, it's getting quite hard to read.

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-sep11

File: src/crimson/osd/pg.cc [hunk 4]
```diff
@@ -1089,37 +1085,34 @@ PG::do_osd_ops_execute(
...
-        }
+    rollbacker.rollback_obc_if_modified(e);
+    // record error log
+    auto maybe_submit_error_log =
+      seastar::make_ready_future<std::optional<eversion_t>>(std::nullopt);
+    // call submit_error_log only for non-internal clients
+    if constexpr (!std::is_same_v<Ret, void>) {
+      if(op_info.may_write()) {
+        maybe_submit_error_log =
+          submit_error_log(m, op_info, obc, e, rep_tid);
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc]: https://pulpito.ceph.com/matan-2024-09-11_13:53:46-crimson-rados-wip-matanb-crimson-testing-sep11-distro-crimson-smithi/

Seems from testing that the obc may have not actually rollbacked here.
Both only failures are `TestClsRbd.directory_methods`.
We had already failed this test in the past and it was fixed here - https://github.com/ceph/ceph/pull/54931/commits/999845c25121cf8d15fdce6b23ee5b3c4b584a9b.

Comment (Matan-B) [reply src/crimson/osd/pg.cc]: Perhaps the `obc` submitted in l.1096 is not actually referencing the one that was rollbacked? 
Will look into this tomorrow.

**Review by Matan-B — CHANGES_REQUESTED** [review]
Related test failure, left comment.

Note: It seems like it did solve the original issue of this PR as all Seastore tests passed.
CC: @cyx1231st

File: src/crimson/osd/pg.cc [hunk 5]
```diff
@@ -975,9 +984,9 @@ PG::do_osd_ops_execute(
   FailureFunc&& failure_func)
 {
   assert(ox);
-  auto rollbacker = ox->create_rollbacker([this] (auto& obc) {
-    return obc_loader.reload_obc(obc).handle_error_interruptible(
-      load_obc_ertr::assert_all{"can't live with object state messed up"});
+  auto rollbacker = ox->create_rollbacker(
+    [object_context=duplicate_obc(obc)] (auto& obc) {
+    obc = duplicate_obc(object_context);
```
Comment (cyx1231st) [inline src/crimson/osd/pg.cc:989]: Why `obc` needs to be duplicated twice (988 and 989)?  Is `duplicate_obc() unnecessary at 989 ?

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:989]: The first copy happens before `OpsExecuter::execute_op()` and serves as a backup of the original obc, while the second during the rollback and recovers the original obc.

Comment (cyx1231st) [reply src/crimson/osd/pg.cc:989]: What can be wrong if the obc is only copied at the `do_osd_ops_execute()` time ?

File: src/crimson/osd/pg.cc (hunk 4 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: > Perhaps the `obc` submitted in l.1096 is not actually referencing the one that was rollbacked? Will look into this tomorrow.

Yes, I think so, the following change should fix the issue:
```
diff --git a/src/crimson/osd/pg.cc b/src/crimson/osd/pg.cc
index f8473daecca..d100775173c 100644
--- a/src/crimson/osd/pg.cc
+++ b/src/crimson/osd/pg.cc
@@ -1081,8 +1081,7 @@ PG::do_osd_ops_execute(
       std::move(all_completed_fut)
     );
   }, OpsExecuter::osd_op_errorator::all_same_way(
-    [this, op_info, m, obc,
-     rollbacker, failure_func_ptr]
+    [this, op_info, m, rollbacker, failure_func_ptr]
     (const std::error_code& e) mutable {
     ceph_tid_t rep_tid = shard_services.get_tid();
     rollbacker.rollback_obc_if_modified(e);
@@ -1093,7 +1092,7 @@ PG::do_osd_ops_execute(
     if constexpr (!std::is_same_v<Ret, void>) {
       if(op_info.may_write()) {
         maybe_submit_error_log =
-          submit_error_log(m, op_info, obc, e, rep_tid);
+          submit_error_log(m, op_info, rollbacker.get_obc(), e, rep_tid);
       }
     }
     return maybe_submit_error_log.then_interruptible(

```

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: Just fixed the code, please test this PR again.

File: src/crimson/osd/pg.cc (hunk 5 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:989]: > What can be wrong if the obc is only copied at the `do_osd_ops_execute()` time ?

Sorry, I'm not following, could you elaborate about this a bit?

Comment (cyx1231st) [reply src/crimson/osd/pg.cc:989]: > https://github.com/ceph/ceph/pull/59543#discussion_r1756169173

What can go wrong if:
```c++
auto rollbacker = ox->create_rollbacker(
    [object_context=duplicate_obc(obc)] (auto& obc) {
    obc = object_context;
});
```
?

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:989]: Ah:-D, yeah, changing

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:989]: @Matan-B I think the problem is actually here, when rolling back, we were not updating the one in the obc_registry but `OpsExecuter::obc`, will correct it.

**Comment by xxhdx1985126** [comment]
@Matan-B @cyx1231st  just fixed the teuthology issue, please take a look again, thanks:-)

File: src/crimson/osd/pg.cc [hunk 6]
```diff
@@ -1088,28 +1088,27 @@ PG::do_osd_ops_execute(
...
     // record error log
     auto maybe_submit_error_log =
-      seastar::make_ready_future<std::optional<eversion_t>>(std::nullopt);
+      interruptor::make_ready_future<std::optional<eversion_t>>(std::nullopt);
     // call submit_error_log only for non-internal clients
     if constexpr (!std::is_same_v<Ret, void>) {
       if(op_info.may_write()) {
         maybe_submit_error_log =
-          submit_error_log(m, op_info, obc, e, rep_tid);
+          submit_error_log(m, op_info, rollbacker.get_obc(), e, rep_tid);
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc:1102]: This change doesn't belong to this commit

File: src/crimson/osd/pg.cc [hunk 7]
```diff
@@ -1064,61 +1079,54 @@ PG::do_osd_ops_execute(
           ceph_assert(e.value() == EDQUOT ||
                       e.value() == ENOSPC ||
                       e.value() == EAGAIN);
-          return rollbacker.rollback_obc_if_modified(e).then_interruptible(
-          [e, failure_func_ptr] {
-            // no need to record error log
-            return (*failure_func_ptr)(e);
-          });
+          rollbacker.rollback_obc_if_modified(e);
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:1082]: Yeah, will do.

File: src/crimson/osd/pg.cc [hunk 8]
```diff
@@ -975,9 +984,9 @@ PG::do_osd_ops_execute(
   FailureFunc&& failure_func)
 {
   assert(ox);
-  auto rollbacker = ox->create_rollbacker([this] (auto& obc) {
-    return obc_loader.reload_obc(obc).handle_error_interruptible(
-      load_obc_ertr::assert_all{"can't live with object state messed up"});
+  auto rollbacker = ox->create_rollbacker(
+    [object_context=duplicate_obc(obc)] (auto& obc) mutable {
+    obc->update_from(*object_context);
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc:989]: @rzarzynski shared some concerns with me regarding avoiding the reload here. We'll need to verify that the actual obc stored in the registry (`obc_registry`) will be updated accordingly so there won't be any inconsistencies between instances of the same obc.
I'm approving in order to schedule testing, we should probably have another look before merging though.

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-59543/

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-09-15_11:59:21-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-09-16_07:15:55-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/

@xxhdx1985126, Bluestore failed with:
```
ERROR 2024-09-16 08:00:01,693 [shard 2:main] none - /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-4879-g81fedcb3/rpm/el9/BUILD/ceph-19.3.0-4879-g81fedcb3/src/crimson/net/io_handler.h:330 : In function 'bool crimson::net::IOHandler::shard_states_t::try_enter_out_dispatching()', abort(%s)
 3# ceph::__ceph_abortf(char const*, int, char const*, char const*, ...) in ceph-osd
 4# crimson::net::IOHandler::shard_states_t::try_enter_out_dispatching() in ceph-osd
 5# crimson::net::IOHandler::notify_out_dispatch() in ceph-osd
 6# crimson::net::IOHandler::do_send(unsigned long, unsigned int, seastar::foreign_ptr<std::unique_ptr<Message, crimson::common::UniquePtrDeleter> >) in ceph-osd
 7# crimson::net::IOHandler::send_recheck_shard(unsigned long, unsigned int, seastar::foreign_ptr<std::unique_ptr<Message, crimson::common::UniquePtrDeleter> >) in ceph-osd
```

To me that looks unrelated as all other thrashers passed.

For Seastore,  2 failures:
 `LibRadosAio.PoolEIOFlag` which is a known failure.
 `TestImageWatcher.NotifySnapProtect` looks new and might be related. 

Let me know what do you think as well. I'll schedule a rerun in the meanwhile.

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
> https://pulpito.ceph.com/matan-2024-09-16_07:15:55-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/
> 
> @xxhdx1985126, Bluestore failed with:
> 
> ```
> ERROR 2024-09-16 08:00:01,693 [shard 2:main] none - /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-4879-g81fedcb3/rpm/el9/BUILD/ceph-19.3.0-4879-g81fedcb3/src/crimson/net/io_handler.h:330 : In function 'bool crimson::net::IOHandler::shard_states_t::try_enter_out_dispatching()', abort(%s)
>  3# ceph::__ceph_abortf(char const*, int, char const*, char const*, ...) in ceph-osd
>  4# crimson::net::IOHandler::shard_states_t::try_enter_out_dispatching() in ceph-osd
>  5# crimson::net::IOHandler::notify_out_dispatch() in ceph-osd
>  6# crimson::net::IOHandler::do_send(unsigned long, unsigned int, seastar::foreign_ptr<std::unique_ptr<Message, crimson::common::UniquePtrDeleter> >) in ceph-osd
>  7# crimson::net::IOHandler::send_recheck_shard(unsigned long, unsigned int, seastar::foreign_ptr<std::unique_ptr<Message, crimson::common::UniquePtrDeleter> >) in ceph-osd
> ```
> 
> To me that looks unrelated as all other thrashers passed. I'll open a tracker once we merge this one.
> 
> For Seastore, 2 failures: `LibRadosAio.PoolEIOFlag` which is a known failure. `TestImageWatcher.NotifySnapProtect` looks new and might be related.
> 
> Let me know what do you think as well. I'll schedule a rerun in the meanwhile.
> 
> -- Edit: Rerun looks ok, all seastore passed Bluestore failed w/ 2 infra and one known. I'm ok with merging this one based on the latest tests. https://pulpito.ceph.com/matan-2024-09-16_16:23:48-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/
> 
> [7907373](https://pulpito.ceph.com/matan-2024-09-16_16:23:48-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/7907373) - infra [7907374](https://pulpito.ceph.com/matan-2024-09-16_16:23:48-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/7907374) - infra [7907382](https://pulpito.ceph.com/matan-2024-09-16_16:23:48-crimson-rados-wip-matanb-crimson-testing-59543-distro-crimson-smithi/7907382) - https://tracker.ceph.com/issues/67568

I'm ok with merging this PR, too.

**Comment by Matan-B** [comment]
Merging, opened a follow up tracker regarding in-flight obc management https://tracker.ceph.com/issues/68130
CC: @rzarzynski, @athanatos
